linera-service/tests/end_to_end_tests.rs

```diff
@@ -1581,145 +1657,395 @@ async fn test_wasm_end_to_end_amm(config: impl LineraNetConfig) -> Result<()> {
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::from_tokens(200)),
+        ])
+        .await;
+    app_fungible1_amm
+        .assert_balances([
+            (owner0, Amount::ZERO),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::from_tokens(128)),
+        ])
         .await;
 
-    app_fungible0_admin
+    app_fungible0_0
         .assert_balances([
             (owner0, Amount::ZERO),
-            (owner1, Amount::from_tokens(62)),
-            (owner_admin, Amount::ZERO),
-            (owner_amm, Amount::from_tokens(188)),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
         ])
         .await;
-    app_fungible1_admin
+    app_fungible1_0
         .assert_balances([
             (owner0, Amount::from_tokens(50)),
-            (owner1, Amount::from_millis(79_680)),
-            (owner_admin, Amount::ZERO),
-            (owner_amm, Amount::from_millis(120_320)),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
         ])
         .await;
 
-    app_amm1.swap(owner1, 0, Amount::from_tokens(50)).await;
-    node_service_admin.process_inbox(&chain_admin).await?;
+    app_fungible0_1
+        .assert_balances([
+            (owner0, Amount::ZERO),
+            (owner1, Amount::from_tokens(70)),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
+        ])
+        .await;
+    app_fungible1_1
+        .assert_balances([
+            (owner0, Amount::ZERO),
+            (owner1, Amount::from_tokens(72)),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
+        ])
+        .await;
 
-    app_fungible0_admin
+    app_amm1.swap(owner1, 0, Amount::from_tokens(50)).await?;
+    node_service_amm.process_inbox(&chain_amm).await?;
+    node_service1.process_inbox(&chain1).await?;
+
+    app_fungible0_amm
         .assert_balances([
             (owner0, Amount::ZERO),
-            (owner1, Amount::from_tokens(12)),
-            (owner_admin, Amount::ZERO),
-            (owner_amm, Amount::from_tokens(238)),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::from_tokens(250)),
         ])
         .await;
-    app_fungible1_admin
+    app_fungible1_amm
+        .assert_balances([
+            (owner0, Amount::ZERO),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::from_millis(102_400)),
+        ])
+        .await;
+
+    app_fungible0_0
+        .assert_balances([
+            (owner0, Amount::ZERO),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
```

Comment:
All those ZERO are annoying.
Could we have a function that calls download all the users and if the entry is missing in the passed array, then the value should be ZERO.